common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java

```diff
@@ -243,6 +247,114 @@ public static void load(String name, ClassLoader loader) {
+    }
+
+    /**
+     * Try to load the helper Class into specified ClassLoader.
+     * @param loader - The specified ClassLoader
+     * @param helper - The helper Class
+     * @return A new helper Class defined in the specified ClassLoader.
+     * @throws ClassNotFoundException Helper class not found or loading failed
+     */
+    private static Class<?> tryToLoadClass(final ClassLoader loader, final Class<?> helper)
+            throws ClassNotFoundException {
+        try {
+            return loader.loadClass(helper.getName());
+        } catch (ClassNotFoundException e) {
+            // The helper class NOT found in target ClassLoader, we have to define the helper class.
+            final byte[] classBinary = classToByteArray(helper);
+            return AccessController.doPrivileged(new PrivilegedAction<Class<?>>() {
+                @Override
+                public Class<?> run() {
+                    try {
+                        // Define the helper class in the target ClassLoader,
+                        //  then we can call the helper to load the native library.
+                        Method defineClass = ClassLoader.class.getDeclaredMethod("defineClass", String.class,
+                                byte[].class, int.class, int.class);
+                        defineClass.setAccessible(true);
+                        return (Class<?>) defineClass.invoke(loader, helper.getName(), classBinary, 0,
+                                classBinary.length);
+                    } catch (Exception e) {
+                        throw new IllegalArgumentException("Define class failed!", e);
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * Load the helper Class as a byte array, to be redefined in specified ClassLoader.
+     * @param clazz - The helper Class provided by this bundle
+     * @return The binary content of helper class.
+     * @throws ClassNotFoundException Helper class not found or loading failed
+     */
+    private static byte[] classToByteArray(Class<?> clazz) throws ClassNotFoundException {
+        String fileName = clazz.getName();
+        int lastDot = fileName.lastIndexOf('.');
+        if (lastDot > 0) {
+            fileName = fileName.substring(lastDot + 1);
+        }
+        URL classUrl = clazz.getResource(fileName + ".class");
+        if (classUrl == null) {
+            throw new ClassNotFoundException(clazz.getName());
+        }
+        try {
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            InputStream in = classUrl.openStream();
+            try {
+                byte[] buf = new byte[1024];
+                while (true) {
+                    int r = in.read(buf);
+                    if (r == -1) {
+                        break;
+                    }
+                    out.write(buf, 0, r);
+                }
+                return out.toByteArray();
+            } finally {
+                in.close();
```

Comment:
put the `close()` calls into a `try { } catch (IOException ignore)` block so we not swallow other exceptions.